### PR TITLE
planner/cascades: add transformation rule TransformAggToProj

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve? 
This PR adds Transformation rule TransformAggToProj in cascades planner(#13709).

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -1952,3 +1953,56 @@ func (r *TransformAggregateCaseToSelection) isTwoOrThreeArgCase(expr expression.
...
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to convert agg to proj.
+func (r *TransformAggToProj) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	agg := old.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	if ok, proj := plannercore.ConvertAggToProj(agg, old.GetExpr().Schema()); ok {
+		newProjExpr := memo.NewGroupExpr(proj)
+		newProjExpr.SetChildren(old.GetExpr().Children...)
+		return []*memo.GroupExpr{newProjExpr}, true, true, nil
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: Only need to `eraseOld` here?
IMHO, currently `eraseAll` is only used when generating `TableDual`?

File: planner/core/rule_aggregation_push_down.go [hunk 2]
```diff
@@ -305,9 +305,10 @@ func (a *aggregationPushDownSolver) pushAggCrossUnion(agg *LogicalAggregation, u
 	// this will cause error during executor phase.
 	for _, key := range unionChild.Schema().Keys {
 		if tmpSchema.ColumnsIndices(key) != nil {
-			proj := a.convertAggToProj(newAgg)
-			proj.SetChildren(unionChild)
-			return proj
+			if ok, proj := ConvertAggToProj(newAgg, newAgg.schema); ok {
```
Comment (francis0407) [inline planner/core/rule_aggregation_push_down.go]: If it cannot be converted, we can break the loop here?

Comment (SeaRise) [reply planner/core/rule_aggregation_push_down.go]: yes, `ConvertAggToProj` will always be false in the loop

File: planner/cascades/testdata/integration_suite_out.json [hunk 3]
```diff
@@ -224,12 +222,10 @@
...
         "SQL": "select sum(a) from t group by a, a+b order by a",
         "Plan": [
-          "Projection_10 8000.00 root Column#3",
-          "└─Sort_20 8000.00 root test.t.a:asc",
-          "  └─HashAgg_13 8000.00 root group by:Column#10, Column#11, funcs:sum(Column#8)->Column#3, funcs:firstrow(Column#9)->test.t.a",
-          "    └─Projection_16 10000.00 root cast(test.t.a)->Column#8, test.t.a, test.t.a, plus(test.t.a, test.t.b)->Column#11",
-          "      └─TableReader_14 10000.00 root data:TableScan_15",
-          "        └─TableScan_15 10000.00 cop[tikv] table:t, range:[-inf,+inf], keep order:false, stats:pseudo"
+          "Projection_13 10000.00 root Column#3",
+          "└─Projection_15 10000.00 root cast(test.t.a)->Column#3, test.t.a",
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply planner/cascades/testdata/integration_suite_out.json]: ok

**Comment by SeaRise** [comment]
@francis0407 
As for `select sum(a) from t group by a, a+b order by a`.
`onPhaseExploration` will create groups:
```
"Group#0 Schema:[Column#14]",
"    Projection_5 input:[Group#1], Column#13",
"Group#1 Schema:[Column#13,test.t.a]",
"    Sort_4 input:[Group#2], test.t.a:asc",
"Group#2 Schema:[Column#13,test.t.a]",
"    Projection_6 input:[Group#3], cast(test.t.a, decimal(65,0) BINARY)->Column#13, test.t.a",
"Group#3 Schema:[test.t.a,test.t.b], UniqueKey:[test.t.a]",
"    TiKVSingleGather_9 input:[Group#4], table:t",
"Group#4 Schema:[test.t.a,test.t.b]",
"    TableScan_8 table:t, pk col:test.t.a"
```
And In the `onPhaseImplementation`, Sort_4 will be implemented to `NominalSortImpl`.
`NominalSortImpl.AttachChildren = children[0]`.
So `Projection_5` and `Projection_6` can not merge.

IMHO, we can implement a Enforcer like `MergeAdjacentProjection`.
If you think it is ok, I will implement a enforcer that `MergeAdjacentProjectionEnforcer` in a new pr.

**Comment by francis0407** [comment]
@SeaRise, good catch. Thanks for the investigation. I think the solution needs some furthur discussion and it is not urgent. Maybe it could be solved by `InlineProjection`.

**Comment by francis0407** [comment]
PTAL @winoros

**Comment by zz-jason** [comment]
Hi @SeaRise this PR is conflicted with the master branch, please resolve the conflicts if free.

**Pull request merged**